### PR TITLE
chore: es6 module

### PR DESCRIPTION
**Pull request by prestwich**

### CODE REVIEW

File: js/lib/BTCUtils.js [hunk 1]
```diff
@@ -1,640 +1,770 @@
...
- * Copyright (c) 2019
  *
- * long description for the file
+ * @file Part of the [bitcoin-spv]{@link https://github.com/summa-tx/bitcoin-spv} project
  *
- * @title BitcoinSPV
- * @summary short description for the file
+ * @title BTCUtils
+ * @summary Utility functions for parsing Bitcoin datastructures
  * @author James Prestwich <jamese@summa.one>
```
Comment (ErinHales) [inline js/lib/BTCUtils.js]: Is your email really `jamese@summa.one` or is it just `james@summa.one`?

**Comment by ErinHales** [comment]
Have you run the tests?  I am getting errors when I try to run them, almost like it doesn't recognize the es6 syntax?

**Comment by prestwich** [comment]
Yeah, make sure you install the new dev dependency

**Comment by ErinHales** [comment]
Hm, I ran `npm i` already...

**Comment by prestwich** [comment]
tests are run by going into the `solidity/` or `js/` directories, and running `npm run test`

File: js/src/ValidateSPV.js [hunk 2]
```diff
@@ -1,232 +1,253 @@
...
-  /**
-   * @notice                Checks validity of header work
-   * @param {Uint8Array}    digest Header digest
-   * @param {Uint8Array}    target The target threshold
-   * @returns {Boolean}     True if header work is valid, false otherwise
-   */
-  validateHeaderWork: (digest, target) => {
-    if (digest === 0) {
-      return false;
+  // Initialize header start index
```
Comment (ErinHales) [inline js/src/ValidateSPV.js]: Don't need this line

File: js/src/ValidateSPV.js [hunk 3]
```diff
@@ -1,232 +1,253 @@
...
+  let totalDifficulty = BigInt(0);
+
+  for (let i = 0; i < headers.length / 80; i += 1) {
+    // ith header start index and ith header
+    const start = i * 80;
+    const header = utils.safeSlice(headers, start, start + 80);
+
+    // After the first header, check that headers are in a chain
+    if (i !== 0) {
+      if (!module.exports.validateHeaderPrevHash(header, digest)) {
```
Comment (dominiquejane) [inline js/src/ValidateSPV.js]: don't need 'module.exports' as long as the function is defined above, so `validateHeaderPrevHash` needs to be moved.

File: js/src/ValidateSPV.js [hunk 4]
```diff
@@ -1,232 +1,253 @@
...
-    // Compare prevHash of current header to previous header's digest
-    if (!utils.typedArraysAreEqual(prevHash, prevHeaderDigest)) {
-      return false;
+
+    // ith header target
+    const target = BTCUtils.extractTarget(header);
+
+    // Require that the header has sufficient work
+    digest = BTCUtils.hash256(header);
+    if (!module.exports.validateHeaderWork(BTCUtils.reverseEndianness(digest), target)) {
```
Comment (dominiquejane) [inline js/src/ValidateSPV.js]: don't need 'module.exports' as long as the function is defined above, so `validateHeaderWork` needs to be moved.

**Comment by dominiquejane** [comment]
`./js/src/utils/utils.js` didn't get touched by the es6 module fairy

File: js/src/BTCUtils.js [hunk 5]
```diff
@@ -1,640 +1,772 @@
...
+  return utils.safeSlice(header, 36, 68);
+}
+
+/**
+ *
+ * Extracts the transaction merkle root from a block header
+ *
+ * @dev                   Use verifyMerkle to verify proofs with this root
+ * @param {Uint8Array}    header An 80-byte Bitcoin header
+ * @returns {number}      The serialized merkle root (big-endian)
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: This returns Uint8Array

File: js/src/BTCUtils.js [hunk 6]
```diff
@@ -1,640 +1,772 @@
...
+  const leSeqence = extractSequenceLELegacy(input);
+  const beSequence = reverseEndianness(leSeqence);
+  return utils.bytesToUint(beSequence);
+}
+
+/**
+ *
+ * Extracts the VarInt-prepended scriptSig from the input in a tx
+ *
+ * @dev                   Will return hex"00" if passed a witness input
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: `Will return [0x00] if passed a witness input`?

File: js/src/BTCUtils.js [hunk 7]
```diff
@@ -1,640 +1,772 @@
...
+  return utils.safeSlice(vinArr, offset, offset + len);
+}
+
+/**
+ *
+ * Determines whether an input is legacy
+ *
+ * @dev                   False if no scriptSig, otherwise True
+ * @param {Uint8Array}    input The input
+ * @returns {boolean}     True for legacy, False for witness
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Other places, `LEGACY` and `WITNESS` are capitalized.  Should do the same here.

File: js/src/BTCUtils.js [hunk 8]
```diff
@@ -1,640 +1,772 @@
...
+
+/* ************ */
+/* Legacy Input */
+/* ************ */
+
+/**
+ *
+ * Determines the length of a scriptSig in an input
+ *
+ * @dev                   Will return 0 if passed a witness input
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Capitalize WITNESS

File: js/src/BTCUtils.js [hunk 9]
```diff
@@ -1,640 +1,772 @@
...
+/* Output */
+/* ****** */
+
+/**
+ *
+ * Determines the length of an output
+ *
+ * @dev                   5 types: WPKH, WSH, PKH, SH, and OP_RETURN
+ * @param {Uint8Array}    output The output
+ * @returns {number}      The length indicated by the prefix, error if invalid length
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Returns BigInt

File: js/src/BTCUtils.js [hunk 10]
```diff
@@ -1,640 +1,772 @@
...
+  return BigInt(len) + BigInt(8 + 1); // 8 byte value, 1 byte for len itself
+}
+
+/**
+ *
+ * Extracts the output at a given index in the TxIns vector
+ *
+ * @dev                   Iterates over the vout. If you need to extract multiple,
+ *                        write a custom function
+ * @param {Uint8Array}    vout The _vout to extract from
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Remove underscore on `_vout`

File: js/src/BTCUtils.js [hunk 11]
```diff
@@ -1,640 +1,772 @@
...
-      // Grab the next input and determine its length.
-      // Increase the offset by that much
-      offset += module.exports.determineInputLength(utils.safeSlice(vin, offset));
+/**
+ *
+ * Extracts the data from an op return output
+ *
+ * @dev                   Errors if no data or not an op return
+ * @param {Uint8Array}    output The output
+ * @returns {Uint8Array}  Any data contained in the opreturn output, null if not an op return
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Take out `, null if not an op return`.  This is handled with an error now.

File: js/src/BTCUtils.js [hunk 12]
```diff
@@ -1,640 +1,772 @@
...
-      if (offset > vLength) {
-        return false;
-      }
+/**
+ *
+ * Extracts the hash from the output script
+ *
+ * @dev                   Determines type by the length prefix and validates format
+ * @param {Uint8Array}    output The output
+ * @returns {Uint8Array}  The hash committed to by the pk_script, or null for errors
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Take out `, or null for errors`

File: js/src/BTCUtils.js [hunk 13]
```diff
@@ -1,640 +1,772 @@
...
 
-    for (let i = 0; i < nOuts; i += 1) {
-      // Grab the next input and determine its length.
-      // Increase the offset by that much
-      offset += module.exports.determineOutputLength(utils.safeSlice(vout, offset));
+  /* P2SH */
+  if (utils.typedArraysAreEqual(tag, new Uint8Array([0x17, 0xa9, 0x14]))) {
+    // Check for maliciously formatted p2sh
+    if (output[output.length - 1] !== 0x87) {
+      // return null;
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Delete this comment

File: js/src/BTCUtils.js [hunk 14]
```diff
@@ -1,640 +1,772 @@
...
+    offset += determineOutputLength(utils.safeSlice(vout, offset));
+  }
 
-    const root = utils.safeSlice(proof, (proofLength - 32), proofLength);
-    let current = utils.safeSlice(proof, 0, 32);
-    const length = (proofLength / 32) - 1;
+  // Returns false if we jump past the end
+  if (offset > vLength) {
+    return false;
+  }
```
Comment (ErinHales) [inline js/src/BTCUtils.js]: Doesn't this do the same thing as ```return offset === vLength;``` on the line below?  Or maybe it's supposed to go inside the for loop like in `validateVin`?

File: js/src/ValidateSPV.js [hunk 15]
```diff
@@ -1,232 +1,252 @@
...
-   * @param {Uint8Array}    vout Raw bytes length-prefixed output vector
-   * @param {Uint8Array}    locktime 4-byte tx locktime
-   * @returns {Uint8Array}  32-byte transaction id, little endian
-   */
-  calculateTxId: (version, vin, vout, locktime) => btcUtils.hash256(
+/**
+ *
+ * Hashes transaction to get txid
+ *
+ * @dev                   Supports Legacy and Witness
```
Comment (ErinHales) [inline js/src/ValidateSPV.js]: Capitalize `LEGACY` and `WITNESS`

File: js/src/ValidateSPV.js [hunk 16]
```diff
@@ -1,232 +1,252 @@
...
+ *
+ * Parses a tx input from raw input bytes
+ *
+ * @dev                   Supports Legacy and Witness inputs
+ * @param {Uint8Array}    input bytes tx input
+ * @returns {object}      Tx input, sequence number, tx hash, and index
+ */
+export function parseInput(input) {
+  // NB: If the scriptsig is exactly 00, we are witness.
+  // Otherwise we are compatibility or legacy
```
Comment (ErinHales) [inline js/src/ValidateSPV.js]: Capitalize `WITNESS`, `COMPATIBILITY`, and `LEGACY`

File: js/src/ValidateSPV.js [hunk 17]
```diff
@@ -1,232 +1,252 @@
...
+  return {
+    sequence, inputId, inputIndex, inputType
+  };
+}
+
+/**
+ *
+ * Parses a tx output from raw output bytes
+ *
+ * @dev                   Differentiates by output script prefix, handles legacy and witness
```
Comment (ErinHales) [inline js/src/ValidateSPV.js]: Capitalize `LEGACY` and `WITNESS`

File: js/src/ValidateSPV.js [hunk 18]
```diff
@@ -1,232 +1,252 @@
...
+
+/**
+ *
+ * Checks validity of header chain
+ *
+ * @dev                   Compares the hash of each header to the prevHash in the next header
+ * @param {Uint8Array}    headers Raw byte array of header chain
+ * @returns {BigInt}      The total accumulated difficulty of the header chain, or an error code
+ * @throws {TypeError}    When passed a chain that contains junk data
+ * @throws {Error}        When passed an invalid chain, or  header with llow work
```
Comment (ErinHales) [inline js/src/ValidateSPV.js]: Extra `l` in `low`

File: js/utils/utils.js [hunk 19]
```diff
@@ -1,131 +1,139 @@
...
-      start = Number(first);
-    } else {
-      start = first;
-    }
+export function safeSlice(buf, first, last) {
+  let start;
+  let end;
+
+  if (first === null || undefined) { start = 0; }
+  if (last === null || undefined) { end = buf.length - 1; }
```
Comment (ErinHales) [inline js/utils/utils.js]: Remove `- 1`

File: js/src/BTCUtils.js (hunk 14 above)
Comment (prestwich) [reply js/src/BTCUtils.js]: Good catch, it was supposed to be inside

**Pull request merged**